File: merge-two-sorted-lists/mergeTwoLists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    result, current, initialized := &ListNode{}, &ListNode{}, false

    addToResult := func (src *ListNode) {
        if initialized {
            current.Next = src
            current = current.Next
        } else {
            result = src
            current = result
            initialized = true
        }
    }

     for list1 != nil && list2 != nil {
        if(list1.Val < list2.Val) {
            addToResult(list1)
            list1 = list1.Next   
        } else {
            addToResult(list2)
            list2 = list2.Next
        }
    }

    for list1 != nil {
        addToResult(list1)
        list1 = list1.Next   
    }

   for list2 != nil {
        addToResult(list2)
        list2 = list2.Next   
    }

    if(!initialized) {
        return nil
    }

    return result;
}

File: merge-two-sorted-lists/MergeTwoLists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode MergeTwoLists(ListNode list1, ListNode list2)
    {
        ListNode result = null;
        ListNode lastItem = null;
        while (list1 != null && list2 != null)
        {
            if (list1.val < list2.val)
            {
                if (result == null)
                {
                    result = list1;
                    lastItem = list1;
                }
                else
                {
                    lastItem.next = list1;
                    lastItem = lastItem.next;
                }

                list1 = list1.next;
            }
            else
            {
                if (result == null)
                {
                    result = list2;
                    lastItem = list2;
                }
                else
                {
                    lastItem.next = list2;
                    lastItem = lastItem.next;
                }
                list2 = list2.next;
            }
        }

        while (list1 != null)
        {
            if (result == null)
            {
                result = list1;
                lastItem = list1;
            }
            else
            {
                lastItem.next = list1;
                lastItem = lastItem.next;
            }

            list1 = list1.next;
        }

        while (list2 != null)
        {
            if (result == null)
            {
                result = list2;
                lastItem = list2;
            }
            else
            {
                lastItem.next = list2;
                lastItem = lastItem.next;
            }
            list2 = list2.next;
        }

        return result;
    }
}